src/oss-health/HealthTool.cs

```diff
@@ -129,86 +55,96 @@ public HealthTool() : base()
             return null;
         }
 
-        void AppendOutput(OutputBuilder outputBuilder, PackageURL purl, HealthMetrics? healthMetrics)
+        public class Options
         {
-            if (outputBuilder.isTextFormat())
+            [Usage()]
+            public static IEnumerable<Example> Examples
             {
-                outputBuilder.AppendOutput($"Health for {purl} (via {purl})\n");
-                outputBuilder.AppendOutput(healthMetrics?.ToString() ?? string.Empty);
-            }
-            else
-            {
-                outputBuilder.AppendOutput(healthMetrics?.toSarif() ?? Array.Empty<Result>().ToList() );
+                get
+                {
+                    return new List<Example>() {
+                        new Example("Find the source code repository for the given package",
+                        new Options { Targets = new List<string>() {"[options]", "package-url..." } })};
+                }
             }
+
+            [Option('f', "format", Required = false, Default = "text",
+                            HelpText = "selct the output format(text|sarifv1|sarifv2)")]
+            public string? Format { get; set; }
+
+            [Option('o', "output-file", Required = false, Default = "",
+                HelpText = "send the command output to a file instead of stdout")]
+            public string OutputFile { get; set; } = "";
+
+            [Value(0, Required = true,
+                HelpText = "PackgeURL(s) specifier to analyze (required, repeats OK)", Hidden = true)] // capture all targets to analyze
+            public IEnumerable<string>? Targets { get; set; }
         }
 
         /// <summary>
-        /// Parses options for this program.
+        ///     Name of this tool.
+        /// </summary>
+        private const string TOOL_NAME = "oss-health";
+
+        /// <summary>
+        ///     Holds the version string, from the assembly.
         /// </summary>
-        /// <param name="args">arguments (passed in from the user)</param>
-        private void ParseOptions(string[] args)
+        private static readonly string VERSION = typeof(HealthTool).Assembly?.GetName().Version?.ToString() ?? string.Empty;
+
+        private static async Task Main(string[] args)
+        {
+            var healthTool = new HealthTool();
+            await healthTool.ParseOptions<Options>(args).WithParsedAsync(healthTool.RunAsync);
+        }
+
+        private void AppendOutput(IOutputBuilder outputBuilder, PackageURL purl, HealthMetrics? healthMetrics)
```

Comment:
Should callers to this be obligated to check for null and not pass in a null?  The best we can do with null is nothing, so really callers (in this case its a private method so just us) should check nulls and not give nulls because the request to append null to output doesn't make sense.

Reply:
For the IOutputBuilder implementing classes, yes, makes sense., I've changed it..but for this class, since this is just a helper method and not a common one, I rather leave it as it is

Reply:
I don't understand the distinction.

The behavior here isn't consistent in any case, in the text case you output

```csharp
$"Health for {purl} (via {purl})",
                        healthMetrics?.ToString() ?? string.Empty
```

which will give the purl detail along with an empty string.  But for sarif you instead include an empty list

```csharp
outputBuilder.AppendOutput(healthMetrics?.toSarif() ?? Array.Empty<Result>().ToList());
```

I'm not sure we want either of these behaviors.  I think we *don't* want to call this function with nulls.  You should remove the nullability attribute and only call this when the data isn't null.

Reply:
The distinction being that IOutputBuilder is defining the constraint that any output needs to be non null. In the tool, we add the logic according to how the output needs to be - 
In the text case, there is a way to mention that the purl that was given did not generate any health metrics (maybe I should change the string.Empty to a more meaningful message). In the Sarif representation, the absence of the entry denotes the failure in health calculation.

Reply:
> The distinction being that IOutputBuilder is defining the constraint that any output needs to be non null

We have thus set the precendent throughout the code that AppendOutput doesn't accept nulls.

> In the tool, we add the logic according to how the output needs to be -

I don't see this as a particularly compelling justification.  What meaningful data do those nulls give?  Any context about how they were generated is gone.  If you care about fixing the null cases you want debug message when a null is generated.  A null output isn't meaningful to the user which is why we don't attempt to output them elsewhere.

> In the text case, there is a way to mention that the purl that was given did not generate any health metrics (maybe I should change the string.Empty to a more meaningful message).

The empty set is different than null.  Empty set is a valid input, null is not.

>  In the Sarif representation, the absence of the entry denotes the failure in health calculation.

The absence of entry can be accomplished without calling this function.  When I trace through what is going to happen here:

You're effectively calling 

`outputBuilder.AppendOutput(new List<Result>());`

This is "fine" because it matches the signature for those functions, but actually they don't take that because they immediately try to cast.  For the string builder output, it actually takes `IEnumerable<string>` and for the sarif case it actually takes `IEnumerable<SarifResult>`. So actually in either case you're going to get an invalid cast exception.

Please fix this function by removing the nullable annotation and ensuring it does not get called with null.

Reply:
I think the suggestion to print debug message makes sense...as far as I am able to give notice to the user, I am fine marking it as non nullable. I still would've like to abstract that null check within the function that leave it in the main function since its cleaner, but since it is not reused, its fine either way.